components/common-go/watch/file.go

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package watch
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+	"golang.org/x/xerrors"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+)
+
+type fileWatcher struct {
+	onChange func()
+
+	watcher *fsnotify.Watcher
+
+	hash string
+}
+
+func File(ctx context.Context, path string, onChange func()) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return xerrors.Errorf("unexpected error creating file watcher: %w", err)
+	}
+
+	fw := &fileWatcher{
+		watcher:  watcher,
+		onChange: onChange,
+	}
+
+	// initial hash of the file
+	hash, err := hashConfig(path)
+	if err != nil {
+		return xerrors.Errorf("cannot get hash of file %v: %w", path, err)
+	}
+
+	// visible files in a volume are symlinks to files in the writer's data directory.
+	// The files are stored in a hidden timestamped directory which is symlinked to by the data directory.
+	// The timestamped directory and data directory symlink are created in the writer's target dir.
+	// https://pkg.go.dev/k8s.io/kubernetes/pkg/volume/util#AtomicWriter
+	watchDir, _ := filepath.Split(path)
+	err = watcher.Add(watchDir)
+	if err != nil {
+		watcher.Close()
+		return xerrors.Errorf("unexpected error watching file %v: %w", path, err)
+	}
+
+	log.Infof("starting watch of file %v", path)
+
+	fw.hash = hash
+
+	go func() {
+		defer func() {
+			log.WithError(err).Error("stopping file watch")
```

Comment:
Similar to the [discussion in another PR](https://github.com/gitpod-io/gitpod/pull/9973#discussion_r874823408), logs should have first letter capital to be sentences as they are consumed by humans. Errors on the other hand should never be capitalized.

Reply:
fixed